### ISSUE TITLE
panic: `AST indices should never change within the same revision` with missing-override-decorator and Callable

### ISSUE THREAD

**Issue by correctmost** [astral-sh/ty#3664]
### Summary

ty crashes when checking this code:

**ty.toml**
```toml
[rules]
missing-override-decorator = 'error'
```

**c.py**
```python
class C:
    def m(self):
        pass
```

**d.py**
```python
from collections.abc import Callable
from c import C

def wrap(f: _T) -> Callable[[], _T]:
    return f

class D(C):
    @wrap(C.m)
    def m(self):
        pass
```

```
error[panic]: Panicked at crates/ty_python_semantic/src/types/function.rs:397:18 when checking `/home/user/d.py`: `AST indices should never change within the same revision`
```

https://play.ty.dev/10c831c9-2625-4fb0-a7f1-30bf0c3611a5
--> The playground reports the crash when I modify `ty.toml` and then switch back to `d.py`

### Version

astral-sh/ruff@43fefbaf (contains the fix for #3654)

**Comment by MichaReiser** [comment]
Thanks for reporting. 

This should now be fixed (playground no longer crashes), see https://github.com/astral-sh/ruff/pull/25621#issue-comment-box

**Comment by correctmost** [comment]
I still see the crash with 0.0.44, which includes the above fix.

I think there is a separate bug in the playground:
- Open https://play.ty.dev/10c831c9-2625-4fb0-a7f1-30bf0c3611a5
- Switch to `d.py`
  - 2 diagnostics are reported
- Switch to `ty.toml`
- Comment out `missing-override-decorator = 'error'`
- Revert the change
- Switch back to `d.py`
  - `Error: unreachable executed`

**Comment by MichaReiser** [comment]
Indeed, funny that you have to comment out `missing-override-decorator` first

**Comment by MichaReiser** [comment]
@lerebear it looks like there's another case where `missing-override-decorator` associates the wrong file. 

> [check_missing_overrides](https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/src/types/overrides.rs#L1125-L1146) recovers its diagnostic anchor from member.ty. The helper added by #25621 filters property accessors to the subclass scope, but its [default branch](https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/src/types/overrides.rs#L1193-L1237) still accepts foreign FunctionLiteral and BoundMethod values unchanged.

**Comment by lerebear** [comment]
A fix for this is now pending review: https://github.com/astral-sh/ruff/pull/25671.
